torch/_dynamo/utils.py

```diff
@@ -1715,3 +1716,232 @@ def build_checkpoint_variable(**options):
+            # ( left_expr ) + ( right_expr )
+            #   left^^^^^       right^^^^^
+            # -2 since end_lineno is 1-indexed and exclusive
+            cur_lineno = expr.left.end_lineno - 2
+            cur_col = normalize(cur_lineno, expr.left.end_col_offset)
+            cur_lineno, cur_col = next_valid_char(cur_lineno, cur_col)
+
+            # heuristic to find the operator character
+            while (ch := lines[cur_lineno][cur_col]).isspace() or ch in ")\\#":
+                if ch in "\\#":
+                    cur_lineno, cur_col = nextline(cur_lineno, cur_col)
+                else:
+                    cur_lineno, cur_col = increment(cur_lineno, cur_col)
+
+            # binary op is 1 or 2 characters long, on the same line
+            right_anchor = cur_col + 1
+            if (
+                right_anchor < len(lines[cur_lineno])
+                and not (ch := lines[cur_lineno][right_anchor]).isspace()
+                and ch not in "\\#"
+            ):
+                right_anchor += 1
+            # right_anchor can be invalid since it is exclusive
+
+            return _Anchors(cur_lineno, cur_col, cur_lineno, right_anchor)
+        elif isinstance(expr, ast.Subscript):
+            # ast gives locations for value and slice subexpressions, e.g.
+            # ( value_expr ) [ slice_expr ]
+            #   value^^^^^     slice^^^^^
+            # find left bracket
+            left_lineno = expr.value.end_lineno - 2
+            left_col = normalize(left_lineno, expr.value.end_col_offset)
+            left_lineno, left_col = next_valid_char(left_lineno, left_col)
+            while lines[left_lineno][left_col] != "[":
+                left_lineno, left_col = increment(left_lineno, left_col)
+            # find right bracket
+            right_lineno = expr.slice.end_lineno - 2
+            right_col = normalize(right_lineno, expr.slice.end_col_offset)
+            right_lineno, right_col = next_valid_char(right_lineno, right_col)
+            while lines[right_lineno][right_col] != "]":
+                right_lineno, right_col = increment(right_lineno, right_col)
+            right_col += 1
+            return _Anchors(left_lineno, left_col, right_lineno, right_col)
+
+    return None
+
+
+def get_instruction_source_311(code: types.CodeType, inst: dis.Instruction):
+    """
+    Python 3.11+ only. Returns lines of source code (from code object `code`)
+    corresponding to `inst`'s location data, and underlines relevant code to `inst`.
+
+    Example: CALL on `g`:
+    f(g(
+      ^^
+        h(x)))
+        ^^^^^
+
+    We need our own implementation since `format_frame_summary` in
+    Python's `traceback` module doesn't handle multi-line expressions
+    (and their anchor extraction code is not completely correct).
+    """
+    result = ""
+    if inst.positions.lineno is None:
+        return result
+    first_line = linecache.getline(code.co_filename, inst.positions.lineno).rstrip()
```

Comment:
this is in the og, comment: empty lines don't have newlines, this makes it consistent that everything doesn't have newline

Reply:
I got this slightly wrong - getline only returns empty if there is an error, so an empty line will give an empty line. OG doesn't have to handle this case because it only prints one line, and it will only process that line if it is non-empty/non-error. We should keep this rstrip + newline logic to handle the case of errors so that line numbers match up in the anchor calculation.